File: cpp/sorting/quick_sort.cpp
#include "quick_sort.h"
#include "utils/array_utils.h"

#include <iostream>

namespace sorting
{
  void quick_sort(int32_t* array, const uint32_t low, const uint32_t high)
  {
    if (low >= high) { return; }

    const uint32_t p = partition(array, low, high);

    quick_sort(array, low, p);
    quick_sort(array, p + 1, high);
  }

  uint32_t partition(int32_t* array, const uint32_t low, const uint32_t high)
  {
    const uint32_t pivot = array[low];
    uint32_t i = low - 1;
    uint32_t j = high + 1;

    while (true)
    {
      do
      {
        i++;
      } while (array[i] < pivot);

      do
      {
        j--;
      } while (array[j] > pivot);

      if (i >= j)
      {
        return j;
      }

      utils::swap(&array[i], &array[j]);
    }
  }
}

File: python/exercises/sorting/quicksort.py
import sys

def quicksort(array):
    return _quicksort(array, 0, len(array) - 1)

def _quicksort(array, low, high):
    if low >= high:
        return

    p = partition(array, low, high)

    _quicksort(array, low, p)
    _quicksort(array, p + 1, high)

def partition(array, low, high):
    pivot = array[low]
    i = low
    j = high

    while True:
        # find the first item greater than the pivot from the left
        while(array[i] < pivot):
            i += 1

        # find the first item less than the pivot from the right
        while(array[j] > pivot):
            j -= 1

        if i >= j:
            # indices are crossed, so we're done
            return j

        # swap the two items
        temp = array[i]
        array[i] = array[j]
        array[j] = temp

        i += 1
        j -= 1

if __name__ ==  "__main__":
    try:
        values = input("Enter unsorted integers: ").strip().split(" ")
        values = [int(v) for v in values]

        quicksort(values)

        print(" ".join([str(v) for v in values]))
    except EOFError:
        print()
    except KeyboardInterrupt:
        print()
